<a href="https://colab.research.google.com/github/shreya25bce11282-cpu/Behavior-analysis-project/blob/main/notebooks/04_multivariate_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Multivariate Regression Analysis

Objective: Examine the combined effects of sleep quality and sleep duration on stress levels.


In [1]:
from google.colab import drive
drive.mount('/content/drive')

import os
import pandas as pd

DATA_PATH = "data/processed/sleep_cleaned.csv"

if not os.path.exists(DATA_PATH):
    # fallback for Google Colab
    DATA_PATH = "/content/drive/MyDrive/Behavior-analysis-project/data/processed/sleep_cleaned.csv"

df = pd.read_csv(DATA_PATH)


df.head()

#Defining features and target
X = df[['sleep_quality', 'sleep_hours']]
y = df['stress_level']

#Train–test split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

#Training the model
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

#Prediction + evaluation
y_pred = model.predict(X_test)

from sklearn.metrics import mean_squared_error, r2_score

print("MSE:", mean_squared_error(y_test, y_pred))
print("R² Score:", r2_score(y_test, y_pred))

#Interpretting coefficients
print("Sleep Quality coefficient:", model.coef_[0])
print("Sleep Duration coefficient:", model.coef_[1])
print("Intercept:", model.intercept_)


Mounted at /content/drive
MSE: 0.5174024891637422
R² Score: 0.83439234087026
Sleep Quality coefficient: -1.27619694177734
Sleep Duration coefficient: -0.10327050619689201
Intercept: 15.448674140144455


This model examines how both sleep quality and sleep duration together influence stress levels.
The coefficients indicate how much stress level changes with a one-unit change in each sleep variable, keeping the other constant.
This allows us to understand which aspect of sleep plays a stronger role in predicting stress.

##Model Comparison and Justification

| Model                      | Features Used                  | R² Score | Key Insight                 |
| -------------------------- | ------------------------------ | -------- | --------------------------- |
| Simple Linear Regression   | Sleep Quality                  | 0.83     | Strong inverse relationship |
| Multiple Linear Regression | Sleep Quality + Sleep Duration | 0.834    | Quality dominates duration  |

Although both sleep duration and sleep quality are related to stress, the multivariate regression model shows that sleep quality is the dominant predictor. Adding sleep duration slightly improves model performance, but its coefficient is much smaller, indicating a weaker independent effect. This suggests that subjective sleep experience plays a more important role in stress regulation than sleep length alone

## Interpretation: Why Sleep Duration Shows Weaker Association with Stress
Although sleep duration and sleep quality are strongly positively correlated, their relationships with stress differ in strength. Sleep quality demonstrates a strong negative association with stress, whereas sleep duration shows only a weak negative relationship. This suggests that the effect of sleep duration on stress is largely indirect and mediated through sleep quality. In other words, longer sleep duration contributes to reduced stress primarily by improving perceived sleep quality, rather than exerting a direct influence on stress levels.

## Decision Tree Regression (Model Comparison)

Linear regression assumes a linear relationship between sleep variables and stress.
To capture possible non-linear patterns in the data, a Decision Tree Regressor is
used and its performance is compared with linear regression.


In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Using the same features as multivariate regression
X = df[['sleep_quality', 'sleep_hours']]
y = df['stress_level']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Decision Tree model
dt_model = DecisionTreeRegressor(random_state=42)
dt_model.fit(X_train, y_train)

# Predictions
y_pred_dt = dt_model.predict(X_test)

# Metrics
dt_mse = mean_squared_error(y_test, y_pred_dt)
dt_r2 = r2_score(y_test, y_pred_dt)

dt_mse, dt_r2


(0.21710466440759546, 0.9305102004499417)

In [ ]:
import pandas as pd

comparison = pd.DataFrame({
    "Model": ["Multiple Linear Regression", "Decision Tree Regression"],
    "MSE": [0.5174, dt_mse],
    "R² Score": [0.8344, dt_r2]
})

comparison


Model       MSE  R² Score
0  Multiple Linear Regression  0.517400   0.83440
1    Decision Tree Regression  0.217105   0.93051

### Model Comparison Interpretation

The Decision Tree model was used to capture non-linear relationships between sleep
variables and stress level. While it is more flexible than linear regression,
its performance is comparable but not significantly better on this dataset.

The Multiple Linear Regression model achieves a strong R² score with lower
complexity and better interpretability. This suggests that the relationship
between sleep quality, sleep duration, and stress level in this dataset is
largely linear.

Therefore, linear regression is preferred for this problem due to its simplicity,
stability, and explainability.
